README.md
# scDD: An R package to identify differentially distributed genes in scRNA-seq 
Version 0.99.0 of R Package 'scDD' (for submission to Bioconductor).  To access 
previous versions of the package (including 1.1.0 as used in the [Genome 
Biology publication]
(https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1077-y), 
see the 
[Releases page](https://github.com/kdkorthauer/scDD/releases)).

To Install:

```
install.packages("devtools")

devtools::install_github("kdkorthauer/scDD")
```

For examples and tips on using the package, please see the vignette PDF, 
which you can bring up by typing 

```
vignette("scDD"")
```

after installing and loading the package into R.  You can also preview the
text and code in the vignette (without the results it produces) in the raw
.Rnw file without first installing scDD by clicking [here](https://github.com/kdkorthauer/scDD/blob/master/vignettes/scDD.Rnw).

Please send bug reports and feature requests by opening a new issue on 
[this page](https://github.com/kdkorthauer/scDD/issues). 

Please cite the following publication if you use scDD in your work:
> Korthauer KD, Chu LF, Newton MA, Li Y, Thomson J, Stewart R, 
Kendziorski C. A statistical approach for identifying differential 
distributions in single-cell RNA-seq experiments. Genome Biology. 
2016 Oct 25;17(1):222. 
[https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1077-y]
(https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-1077-y)

Note: The package was built using the development version of R (3.4.0,
2016-10-26 r71594)


inst/NEWS
Changes in Version 1.13.1 (2020-08-17)
______________________________________

o When testing for difference in proportion of zeroes, if the Bayesian logistic
  regression fails to converge for a given gene (for example, if there is perfect
  separation between conditions), then NA will be returned (instead of throwing 
  an error. 

Changes in Version 1.3.4 (2018-03-26)
______________________________________

o The test for differences in proportion of zeroes now uses the Wald test 
  p-value instead of the likelihood ratio. This will give very similar results,
  but the Wald test is slightly more conservative.

Changes in Version 1.3.3 (2018-03-23)
______________________________________

o An option has been added to skip the categorization step if only intereseted
  significance of difference. This will speed up computation.
  
o The testing zeroes step and the KS test have been parallelized to speed up
  computation.

Changes in Version 1.1.7 (2017-10-23)
_____________________________________

o A parameter 'level' was added to the main 'scDD' function which allows the 
  user to control the significance level used as the cutoff for considering a 
  gene to be differentially distributed. Previously it was fixed at 0.05 (now 
  the default value).

o Two columns were added to the results object that contain (1) an overall 
  combined p-value (via Fisher's method) for nonzero and zero differences, and 
  (2) Benjamini-Hochberg adjusted version of (1).

o If `testZeroes` is FALSE, the columns with p-values for the test of 
  differential dropout are no longer included. Previously they held NA values.

o If `testZeroes` is TRUE, all zero test p-values are included in the output. 
  Previously, only the tests where the nonzero test was not significant were 
  reported.

Changes in Version 1.1.5 (2017-09-27)
_____________________________________

o The input and output object of the main scDD function has been changed from 
  SummarizedExperiment to SingleCellExperiment to increase interoperability 
  among other Bioconductor packages

o The simulateSet function now returns a SingleCellExperiment object 
  (previously it returned an object in list format)

o The preprocess function now takes as input a SingleCellExperiment object 
  instead of a list of data frames

Changes in Version 1.0.0 (2017-04-23)
_____________________________________

o This is the Bioconductor 3.5 Release version

o scDD is a package for identifying differentially distributed genes in 
  single-cell RNA-seq data. It is designed to detect differences in expression
  that are more complex than a simple mean shift, including:

  - traditional differential expression (DE)

  - differential modality (DM)

  - differential proportion of cells in each state (DP)

  - both differential modality and differential proportion (DB)

Changes in Version 0.99.0 (2016-12-07)
______________________________________

o Initial Bioconductor submission version.



R/Preprocess.R
#' preprocess
#'
#' Function to preprocess SingleCellExperiment object (1) to
#' only keep genes with a certain number of nonzero entries,
#' and (2) optionally apply a normalization procedure.
#'
#' @param SCdat An object of class \code{SingleCellExperiment} that contains 
#'  single-cell expression and metadata. The \code{assays} 
#'   slot contains a named list of matrices, where the normalized counts are 
#'   housed in the one named \code{normcounts}, and unnormalized counts are 
#'   stored in the one names \code{counts}. If either \code{scran_norm} or 
#'   \code{median_norm} is set to TRUE, the \code{normcounts} slot will be 
#'   created from the \code{counts} slot. The counts and normalized counts 
#'   matrices should have one
#'    row for each gene and one sample for each column.  
#'   The \code{colData} slot should contain a data.frame with one row per 
#'   sample and columns that contain metadata for each sample.  This data.frame
#'   should contain a variable that represents biological condition, which is 
#'   in the form of numeric values (either 1 or 2) that indicates which 
#'   condition each sample belongs to (in the same order as the columns of 
#'   \code{normcounts}).  Optional additional metadata about each cell can also
#'   be contained in this data.frame, and additional information about the 
#'   experiment can be contained in the \code{metadata} slot as a list.
#' 
#' @param zero.thresh A numeric value between 0 and 1 that represents
#'  the maximum proportion
#'  of zeroes per gene allowable in the processed dataset
#'  
#' @param scran_norm Logical indicating whether or not to normalize the data
#'  using scran Normalization from \code{scran}
#'  
#' @param median_norm Logical indicating whether or not to normalize the data
#'  using Median Normalization from \code{EBSeq}
#'  
#' @param condition A character object that contains the name of the column in 
#' \code{colData} that represents 
#'  the biological group or condition of interest (e.g. treatment versus 
#'  control).  Note that this variable should only contain two 
#'  possible values since \code{scDD} can currently only handle two-group 
#'  comparisons.  The default option assumes that there
#'  is a column named "condition" that contains this variable. 
#' 
#' @export
#' 
#' @importFrom EBSeq MedianNorm
#' 
#' @importFrom EBSeq GetNormalizedMat
#' 
#' @importFrom scran computeSumFactors
#' 
#' @import SummarizedExperiment
#'
#' @import SingleCellExperiment
#' 
#'
#' @references Korthauer KD, Chu LF, Newton MA, Li Y, Thomson J, Stewart R, 
#' Kendziorski C. A statistical approach for identifying differential 
#' distributions
#' in single-cell RNA-seq experiments. Genome Biology. 2016 Oct 25;17(1):222.
#' \url{https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-
#' 1077-y}
#'
#' @return An object of class \code{SingleCellExperiment} with genes removed if
#' they have more than \code{zero.thresh} zeroes, and the \code{normcounts} 
#' assay added if either \code{scran_norm} or \code{median_norm} is set to TRUE
#' and only \code{counts} is provided. If \code{normcounts} already exists and
#' either \code{scran_norm} or \code{median_norm} is set to TRUE, then the new
#' normalized counts are placed in the \code{normcounts} assay slot, and the
#' original values are moved to a new slot called \code{normcounts-orig}.
#'  
#' @examples 
#'  
#'  # load toy example SingleCellExperiment object
#'  
#'  data(scDatEx)
#'  
#'  # apply the preprocess function to filter out genes if they have more than
#'  # 75% zero
#'  
#'  scDatEx <- preprocess(scDatEx, zero.thresh=0.75)
#'  
#'  # apply the preprocess function again, but this time threshold on the 
#'  # proportion of zeroes and apply scran normalization
#'  # set the zero.thresh argument to 0.75 so that genes with more than 75% 
#'  # zeroes are filtered out 
#'  # set the scran_norm argument to TRUE to return scran normalized counts
#'  
#'  scDatEx.scran <- preprocess(scDatEx, zero.thresh=0.75, scran_norm=TRUE)
#'  
#'  # set the median_norm argument to TRUE to return Median normalized counts
#'  
#'  scDatEx.median <- preprocess(scDatEx, zero.thresh=0.75, median_norm=TRUE)

preprocess <- function(SCdat,
                       condition="condition",
                       zero.thresh=0.9,
                       scran_norm=FALSE,
                       median_norm=FALSE){
  
  # check whether SCdat is a member of the SingleCellExperiment class
  if(!("SingleCellExperiment" %in% class(SCdat))){
    stop("Please provide a valid 'SingleCellExperiment' object.")
  }
  
  # check that condition inputs are valid
  if (length(unique(colData(SCdat)[[condition]])) != 2 | 
      length(colData(SCdat)[[condition]]) != dim(SCdat)[2]){
    stop("Error: Please specify valid condition labels.")
  }
  
  if(median_norm & scran_norm){
    stop(paste0("Specified conflicting options for normalization. Only one ",
                " of scran_norm and median_norm can be set to TRUE"))
  }
  
  if (sum(median_norm, scran_norm) == 1 &
        !("counts" %in% assayNames(SCdat)) ){
    stop(paste0("If median or scran norm is specified, the input object ",
                "must contain raw counts in 'counts'."))
  }
  
  if (sum(median_norm, scran_norm) == 1 &
       "normcounts" %in% assayNames(SCdat) ){
    warning(paste0("median or scran norm is specified and the 'normcounts' ",
                "assay already exists; replacing 'normcounts' in output ",
                "with the specified normalization method. Original contents ",
                "of 'normcounts' are now in 'normcounts-orig'."))
    assay(SCdat, "normcounts-orig") <- normcounts(SCdat)
  }
    
  if (sum(median_norm, scran_norm) == 0 &
     !("normcounts" %in% assayNames(SCdat))){
    stop(paste0("If median or scran norm is not specified, the input object ",
                " must contain normalized counts in 'normcounts'"))
  }
  
  
  if (sum(median_norm, scran_norm) == 0){
    pe_mat <- normcounts(SCdat)
  }else{
    pe_mat <- assay(SCdat, "counts")
  }
  
  # reference category/condition - the first listed one
  ref <- unique(colData(SCdat)[[condition]])[1]
  cond <- colData(SCdat)[[condition]]
 
  # threshold on zero percentage (within each condition)
  pz1 <- apply(pe_mat[,cond==ref], 1, function(x) sum(x==0)/length(x))
  pz2 <- apply(pe_mat[,cond!=ref], 1, function(x) sum(x==0)/length(x))
  SCdat <- SCdat[pz1 <= zero.thresh & pz2 <= zero.thresh,]
  
  # normalize
  if (median_norm){ 
    message("Performing Median Normalization")
    libsize <- MedianNorm(assay(SCdat, "counts"))
    normcounts(SCdat) <- GetNormalizedMat(assay(SCdat, "counts"),
                                          libsize)
  }else if(scran_norm){
    message("Performing scran Normalization")
    SCdat <-  computeSumFactors(SCdat, assay.type="counts")
    normcounts(SCdat) <- GetNormalizedMat(assay(SCdat, "counts"),
                                          sizeFactors(SCdat))
  }
  
  return(SCdat)
}
